### ISSUE TITLE
backend_compile appears to leak memory

### ISSUE THREAD

**Issue by DiffeoInvariant** [google/jax#5531]
I'm using jax to evaluate part of a PDE, where the jax function looks something like this:
```
@jax.jit
def evaluate_model(*args):
      inputs, params = args[0], args[1:]
      intermediate_value_1 = func1(*inputs)
      intermediate_value_2 = func2(*inputs)
      intermediate_value_3 = func3(intermediate_value_1,*params)
      return func4(intermediate_value_2,intermediate_value_3)
```
where each of `func1`, `func2`, `func3`, and `func4` have been appropriately `vmap`-ed (i.e. `func1 = jax.vmap(func1,in_axes=my_in_axes)`), because they are written to take data at one quadrature point, but the solver gives the function data at all of the quadrature points in one big array. `func1` and `func2` have also been jit-compiled in advance. I'm also differentiating this function, both with respect to `args[0]` (to solve the PDE) and with respect to `args[1:]` (to solve my inverse problem). The data from the PDE solve (which I'm doing with `fenics`) is converted to a numpy array with `petsc4py`, then converted to a JAX array before being passed to the function, and the output is converted back to numpy and then to `petsc4py` and back into `fenics`.

 My problem is that something in jax is leaking a lot of memory during my optimization loop; I've attached files generated with `jax.profiler.save_device_memory_profile` at various iterations. The leak isn't very big in these files, but over thousands of solves, it adds up quickly. If I'm reading the `pprof` graph correctly, `backend_compile` is the only jax function that's growing, and I've used `tracemalloc` to narrow down the source of the memory growth to `jax/core.py`, so I can be sure that the leak is coming from jax.

[memory_prof.zip](https://github.com/google/jax/files/5882260/memory_prof.zip)

Any idea if my interpretation of the problem (that it's coming from `backend_compile`) is correct? If so, any tips to fix it? I'm happy to provide more profiling data (e.g. during a longer run where the leak gets really big) if that would help.

Note: This seems somewhat similar to https://github.com/google/jax/issues/282, but is sufficiently different that I think it warrants its own issue. Happy to move discussion of this problem to 282 though if this belongs there

EDIT: this zip contains two memory profiles during a longer run to see the problem more clearly. Run `pprof --web --diff_base memory_fwd_0.prof memory_fwd_508.prof` to reproduce the graph
[fwd_diff.zip](https://github.com/google/jax/files/5884198/fwd_diff.zip)

**Comment by jekbradbury** [comment]
Can you enable `config.update('jax_log_compiles', True)` to see if you're repeatedly recompiling (and if so, whether that's unexpected given your setup)?

If you're invoking `evaluate_model` with arbitrarily many different shapes in `*args`, then there will be continued recompilation; in that case, the best solution would probably be for us to evict old GPU binaries from the jit cache.

There's some more discussion, plus workarounds, at https://github.com/google/jax/issues/2072. Closing as a dup of that for now; feel free to reopen (if e.g. your model isn't actually recompiling).

**Comment by DiffeoInvariant** [comment]
Hm, the advice on #2072 seems to not work; I tried re-jitting the function occasionally and that just makes the memory usage grow faster. They reference `jax.interpreters.xla._xla_callable.clear_cache()`, but `jax.interpreters.xla._xla_callable` is a function, and as expected, trying to do `jax.interpreters.xla._xla_callable.clear_cache()` causes an exception to be thrown. Do you know of any other way to clear the XLA cache? I can easily fix the issue at least as far as my code is concerned if I can figure out how to do that.

**Comment by jekbradbury** [comment]
Recompilations even though the shapes are always the same don't sound good! One possibility there is that you're defining `evaluate_model` inside another function, so the Python object `evaluate_model` is different every time that outer function runs. But then the old function objects should be going out of scope and their corresponding executables should be getting garbage collected...

(In fact, I'm not sure there's a way to clear the jit cache right now—and if you need to I'd recommend redefining your callable.)

Basically, I'm not sure what's going on—can you show some of the output of `jax_log_compiles`?

**Comment by DiffeoInvariant** [comment]
Huh, so this is somewhat interesting---occasionally redefining my function (i.e. doing `evaluate_model = jax.jit(evaluate_model)` every set amount of iterations) seems to have changed the behavior completely. Now, the first forward and backward pass gives compilation messages of 
```
WARNING:absl:Compiling unified_scalar_invariant_func for args (AbstractUnit(), AbstractUnit()).
WARNING:absl:Compiling unified_form_invariant_func for args (AbstractUnit(), AbstractUnit()).
WARNING:absl:Compiling _evaluate_jax_cr for args (AbstractUnit(), AbstractUnit(), AbstractUnit(), AbstractUnit()).
WARNING:absl:Compiling _evaluate_jax_cr for args (ShapedArray(float64[108,2,2]), ShapedArray(float64[2]), AbstractUnit(), AbstractUnit()).
WARNING:absl:Compiling backward_pass for args (AbstractUnit(),).
WARNING:absl:Compiling <unnamed wrapped function> for args (ShapedArray(float64[108,2,2]),).
```
which is what I expect on the first run (`_evaluate_jax_cr` is `evaluate_model` in my description above, and the `unified_xxx_invariant_func` functions are the intermediate functions inside of it). Before explicitly re-compiling/re-defining `evaluate_model`, I would get two recompilation warnings from abseil every subsequent iteration, and now I don't get any (i.e. it is no longer recompiling). Also, the memory leak appears to be gone now.

 I don't remember which functions it was recompiling, but if you're interested, I ought to be able to roll my code back a few git commits and reproduce the old, buggy output.


The upshot of this is that I no longer have a problem, as far as I can tell, so feel free to close this issue if you're satisfied with this resolution. It would be nice to understand why explicitly re-defining the function solves the re-compilation issue, but since it works, not necessary from my perspective. Happy to give you whatever info you want/need if you want to look more into it though. FWIW, all of the functions involved here, except one (whose compilation isn't shown here, because it's inside `_evaluate_jax_cr` and gets compiled along with it), are defined inside other functions (and are built dynamically based on runtime information, so it has to be that way)

**Comment by hawkinsp** [comment]
It may or may not be related, but jaxlib 0.1.60 (just released) fixed a memory leak on CPU.
